### PR TITLE
Date reminders

### PR DESCRIPTION
**Pull request by VanAnderson**
Shows upcoming birthdays and start date anniversaries on the admin dashboard. Time interval for upcoming dates is currently set to a week. 

I had to embed the column names directly into the SQL string in the get_upcoming_dates function. However, these are not user inputs and can only be symbols, so I don't think this is vulnerable to SQL injection.

### CODE REVIEW

File: app/views/home/index.html.slim [hunk 1]
```diff
@@ -8,3 +8,9 @@
     = link_to "Roles", admin_roles_path, class: "button"
 -else
   h1 Welcome, Please Sign In
+
+-if current_user && current_user.admin?
```
Comment (jah2488) [inline app/views/home/index.html.slim]: We should use the `admin_user?` helper method that is written in the application controller instead of listing it out by hand.

File: app/views/shared/_upcomming_days.html.slim [hunk 2]
```diff
@@ -0,0 +1,12 @@
+table
+  thead
+    tr
+      th Name
+      th Date
+      th Years
+  tbody
+  -dates.each do |date|
```
Comment (jah2488) [inline app/views/shared/_upcomming_days.html.slim]: There should also be a single space between `-` or `=` and the ruby code that follows.

File: db/migrate/20160620212000_create_engagement_horses.rb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-class CreateSessionHorses < ActiveRecord::Migration
+class CreateEngagementHorses < ActiveRecord::Migration
```
Comment (jah2488) [inline db/migrate/20160620212000_create_engagement_horses.rb]: Yah my bid, thanks for the fix 👍

File: app/controllers/home_controller.rb [hunk 4]
```diff
@@ -1,4 +1,20 @@
...
+    render locals: {
+      birthdays: birthdays,
+      startdays: startdays,
+    }
+  end
+
+  private
+
+  def get_upcoming_dates(type)
+    return unless type.class == Symbol
```
Comment (jah2488) [inline app/controllers/home_controller.rb]: I'm not entirely convinced we need this guard clause here, as long as we are casting the input correctly, it shouldn't cause much of an issue.

In general, in ruby, we avoid during hard type checking and prefer soft 'duck type' checking. 

However, if you're going to do type checking, its preferred to use `type.is_a?(Symbol)` over `==`

File: app/controllers/home_controller.rb [hunk 5]
```diff
@@ -1,4 +1,20 @@
...
+    users = User.where("DATE_PART('year', age(now()+ interval ?, #{type})) - DATE_PART('year', age(#{type})) = 1", '1 week')
```
Comment (jah2488) [inline app/controllers/home_controller.rb]: This SQL query scares me a tiny bit, but it seems to be working just fine. I think we can make this a bit less scary by doing a few things:

1. Use `field` instead of `type` which conveys that what is being given is a class name or a primitive.
1. Lets assign the sql string to a local variable on a separate line, then pass that var into `User.where(sql, "1 week")` That will also allow us to use one of Ruby's multi line strings.
```
sql = <<-SQL
  DATE_PART(
       'year', 
       age(now() + interval ?, #{field})
  ) - DATE_PART(
    'year',
    age(#{field})
) = 1
SQL
```

File: app/controllers/home_controller.rb [hunk 6]
```diff
@@ -1,4 +1,20 @@
...
+    users.map do |user|
+      [user, user.send(type).strftime("%m-%d"), ((Date.today + 7 - user.send(type)).to_i/365.25).floor]
+    end
```
Comment (jah2488) [inline app/controllers/home_controller.rb]: I'm not entirely sure whats going on here, it _looks like_ we're taking the user objects we received and then mapping them into an array that contains the user object itself, the time formatted, and then a week from now - the day of that user's  date divided by days in the week rounded down.

_if_ we are going to go this route and return this extra information from the user, I'd want this in a named object that we can write unit tests against. Something like `UserWithBirthday` or something to that effect, that will take the user and present me with easy method to access what these 3 elements in this array represent. I will want some unit tests for those methods as well.

So from my research it looks like using something from https://www.postgresql.org/docs/8.1/static/functions-datetime.html# is probably the best approach, as you are doing.

However, lets not forget about all the nice date parsing and presenting methods that rails gives us like `beginning_of_week`, `minutes_from_now`, and others. http://apidock.com/rails/ActiveSupport/CoreExtensions/Date/Calculations


So lets create a new object to use for when this method returns its values, it will be an array of that new object that wraps the user object.

Then lets move this method itself into `User` model as a class/singleton method, so it can be called like `User.get_upcoming_dates(:birth_date)`

**Comment by jah2488** [comment]
@VanAnderson To answer your question about SQL injection vulnerabilities, it _should_ be filtered by AR before being run, but I'm not too concerned as this screen will only be accessed by logged in users and we are not using anything from the client in the calculation.

**Comment by VanAnderson** [comment]
@jah2488 

really good feedback! I refactored all of the logic for the upcoming dates into UserWithDates.

It has a class method that takes a field (:start_date, :birth_date) and returns an array of all UserWithDates that have an upcoming anniversary of that field within a give time interval (which can be specified as an integer of days or as a rails duration). 

The age_offset might require some explanation. On the table that displays anniversaries/birthdays, I think it makes the most sense to display how old that user _will be_ on their birthday/start date. I did this by adding 1 to their current age/years with the company. But if today is their actual birthday/anniversary, then you don't want to add 1 because the age has already incremented. So I just add the return value of age_offset instead, which accounts for this. Maybe there is a better way to do this?

            def age_offset(field)
              if is_date?(field)
                0
              else
                1
              end
            end

Although, the current SQL query will not retrieve the current days birthdays/anniversaries, only days in the future. I'm going to write a test for this tomorrow and make it pass, just wanted to make sure I'm on the right track.

File: app/models/user_with_dates.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
+class UserWithDates
+  attr_accessor :user
+  def initialize(user)
+    @user = user
```
Comment (jah2488) [inline app/models/user_with_dates.rb]: You can replace `@user = user` here with `self.user = user`

File: app/models/user_with_dates.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  def get_age(field, from_date = Date.today)
+    ((from_date - @user.send(field)).to_i/365.25).floor
+  end
```
Comment (jah2488) [inline app/models/user_with_dates.rb]: So this method returns their age as a year? Would it be possible to use one of Rail's built in [date calculations](http://api.rubyonrails.org/v2.3/classes/ActiveSupport/CoreExtensions/Date/Calculations.html)? I worry about subtle date bugs appearing at a later date and would rather lean on something built in, if possible.

Since you have `attr_accessor :user` you can replace any situations where you've used `@user` with `self.user` or just simply `user`.

File: app/models/user_with_dates.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    next_annual(@user.send(field))
+  end
+
+  def name
+    @user.name
+  end
+
+  def is_date?(field)
+    next_date(field) == Date.today
+  end
```
Comment (jah2488) [inline app/models/user_with_dates.rb]: I feel this name would be more intent revealing as `today?(field)`

File: app/models/user_with_dates.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  private
+
+  def self.duration_to_integer(time_interval)
+    if time_interval.is_a?(ActiveSupport::Duration)
+      (time_interval.to_i/86400).floor
+    else
+      time_interval
+    end
+  end
```
Comment (jah2488) [inline app/models/user_with_dates.rb]: This method turns something like `7.days` into simply `7` by converting it into seconds and dividing by the amount of seconds in a day. It'd be helpful to have the numbers like that be stored as constants `SECONDS_IN_HOUR`.

File: app/models/user_with_dates.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+  def next_annual(event_date)
+    year = Date.today.year
+    mmdd = event_date.strftime('%m%d')
+    year += 1 if mmdd < Date.today.strftime('%m%d')
+    mmdd = '0301' if mmdd == '0229' && !Date.parse("#{year}0101").leap?
+    return Date.parse("#{year}#{mmdd}")
+  end
```
Comment (jah2488) [inline app/models/user_with_dates.rb]: So there is a lot going on here. From my own understanding, this is getting the month and day from the event_date and then comparing that against the month and day of the current date to increment the year (or not).

Then, if it is not a leap year, and the date is the 29th of Feburary, increment the month/date to be March 1st. Then feed all of that back into a brand new date.

What I would want from this method. Is to pull out some more meaning/responsibilities out of this method into additional private helper methods. 

The loop day predicate should be its own named method to reveal the intent of what is happening.

The "date past today" conditional on line 61 should also be its own predicate method.

We are calling `Date.today` 2 times, and parsing dates from strings 2 times. I feel these could be cleaned up a bit.

We can use a single call to `Date.today` (assigned to a variable), to get the year, the m/d as a string, and to ask if its a leap year.

File: app/models/user_with_dates.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    users.map do |user|
+      UserWithDates.new(user)
+    end
+  end
+
+  private
+
+  def self.duration_to_integer(time_interval)
+    if time_interval.is_a?(ActiveSupport::Duration)
+      (time_interval.to_i/86400).floor
```
Comment (jah2488) [inline app/models/user_with_dates.rb]: Always need a space around operators `(time_interval.to_i / 86400)`

**Comment by jah2488** [comment]
This is looking real good. I left a few comments on the UserWithDates class. The one additional thing the UserWithDates class is missing before this is good to merge is a bunch of unit tests using rspec.

I would go ahead and create a `spec/models/user_with_dates_spec.rb` file and add tests that go through all the different edge cases.

**Comment by VanAnderson** [comment]
Alrighty @jah2488!

  - Accessed necessary instance variables using self.variable_name

  - broke down get_age into parts and tried to make it more readable. I couldn't find a way to integrate the rails date calculations into it from the link you gave, but I think it has improved in readability.

  - renamed is_date? to today?

  - put class constants for SECONDS_IN_DAY and DAYS_IN_YEAR

  - broke down next_annual function into private functions and tried to make it more readable at a glance.

  - wrote another SQL query to also add birthdays/anniversaries that fall on the current day in addition to upcoming birthdays.

  - made sure that the SQL queries are only called in the controller if user is an admin, otherwise it would run the SQL and not use it (since only admins can see birthdays).

  - wrote some RSpec tests for edge cases and general testing. Could totally write more or refactor tests if you think it's necessary!

**Comment by jah2488** [comment]
This is looking great. I'm going to merge this in. 

One note is that we generally want to avoid having if or unless statements inside our tests. I see that was to avoid a leap year issue, but we should instead find another way around that problem. I've included the timecop Ruby gem to help with those testing issues. 

https://github.com/travisjeffery/timecop

**Comment by VanAnderson** [comment]
Gotcha, I might revisit that test and fix it up later. Thanks Justin!
